core/comet/service.go

```diff
@@ -0,0 +1,61 @@
+package comet
+
+import (
+	"context"
+	"time"
+)
+
+type HasBeginBlocker interface {
+	BeginBlock(context.Context, BlockInfo) error
+}
+
+// Service is an interface that can be used to get information specific to Comet
+type Service interface {
+	GetCometInfo(context.Context) BlockInfo
+}
+
+// BlockInfo is the information comet provides apps in ABCI
+type BlockInfo interface {
+	Evidence() []Misbehavior // Evidence misbehavior of the block
+	// ValidatorsHash returns the hash of the validators
+	// For Comet, it is the hash of the next validator set
+	ValidatorsHash() []byte
+	ProposerAddress() []byte       // ProposerAddress returns the address of the block proposer
+	DecidedLastCommit() CommitInfo // DecidedLastCommit returns the last commit info
+}
+
+// MisbehaviorType is the type of misbehavior for a validator
+type MisbehaviorType int32
+
+const (
+	Unknown           MisbehaviorType = 0
+	DuplicateVote     MisbehaviorType = 1
+	LightClientAttack MisbehaviorType = 2
+)
+
+// Validator is the validator information of ABCI
+type Validator interface {
+	Address() []byte
+	Power() int64
+}
+
+// Misbehavior is the misbehavior information of ABCI
+type Misbehavior interface {
+	Type() MisbehaviorType
+	Validator() Validator
+	Height() int64
+	Time() time.Time
+	TotalVotingPower() int64
+}
+
+// CommitInfo is the commit information of ABCI
+type CommitInfo interface {
+	Round() int32
+	Votes() []VoteInfo
+}
+
+// VoteInfo is the vote information of ABCI
+type VoteInfo interface {
+	Validator() Validator
+	SignedLastBlock() bool
+}
```

Comment:
@aaronc why are these all interfaces?

Reply:
Well because in this iteration they're all implemented by the concrete types RequestBeginBlock, RequestPrepareProposal, and RequestProcessProposal. But I'm also not convinced that's the right approach. I think it would be better to have callbacks with either concrete request types or wrapper types like implemented here, as opposed to context like I did in this branch